Autodesk/synthesis
# CAN Support

### Description
Send, receive, and use CAN device data between a robot project and fission.

### Objectives
- [x] Send TalonFX Data
- [x] Send CANSparkMax Data
- [x] Make simulated motor fields appear as CAN devices rather than SimDevices
- [x] CAN Motor Brain Mappings
- [x] CANEncoder support

### Testing Done
- [x] Test TalonFX in WS Viewer
- [x] Test CANSparkMax in WS Viewer
- [x] Make sure CAN Motor inputs map into the brain properly
- [x] Test CANEncoders on SparkMaxes

For testing:
1. Build and publish SyntheSim with `./gradlew build` and `./gradlew publishToMavenLocal`.
2. Build and run the JavaSample with `./gradlew build` and `./gradlew simulateJava`.
3. Run fission
4. Open fission, spawn a robot, then click the WS Test button in the debug menu (this must be done first).
5. Verify the connection by opening the WS Viewer panel, you should see a long list of motors and encoders.
6. Click the RoboRio button to setup CAN output group, select all the CANSparkMax motors and all the WheelDrivers, then click Done
7. Do the same but for Encoders, this time, just select CANSparkMax[1] and any of the wheel stimuli.
8. In the robotSimulation screen, click Autonomous, you should see the robot drive forward 20 (feet?), then stop.

We need to have a discussion about how we want device mapping to work, right now every motor in a group's output is averaged and applied to each driver. We need to make it easier to do one to one mappings between motors and drivers, because right now, clicking through the whole menu ~10 times to make an output group for each motor is ridiculous.

> [!note]
> Pulled from #1077 for it's encoder framework, so I think just turning that PR into a gyro / joystick PR is the way to go.

> [!note]
> CANMotor inputs (eg. supplyCurrent) have been ignored, since they're not simulated in fission.

This PR is a combination of JIRA Issues [1747](https://jira.autodesk.com/browse/AARD-1747), [1748](https://jira.autodesk.com/browse/AARD-1748), and [1749](https://jira.autodesk.com/browse/AARD-1749) for a total of 18 story points.
